changeset/manager.go

```diff
@@ -0,0 +1,432 @@
+			return err
+		}
+
+		if err := change(config, changes); err != nil {
+			return err
+		}
+
+		changeBytes, err := proto.Marshal(changes)
+		if err != nil {
+			return err
+		}
+
+		// ...and update the stored changes
+		changeset.Changes = changeBytes
+		if _, err := m.kv.CheckAndSet(csKey, csVersion, changeset); err != nil {
+			if err == kv.ErrVersionMismatch {
+				// Someone else updated the changes first - try again
+				continue
+			}
+
+			return err
+		}
+
+		return nil
+	}
+}
+
+func (m manager) GetPendingChanges() (int, proto.Message, proto.Message, error) {
+	// Get the current configuration, but don't bother trying to create it if it doesn't exist
+	configVal, err := m.kv.Get(m.key)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+
+	// Retrieve the config data so that we can transform it appropriately
+	config := proto.Clone(m.configType)
+	if err := configVal.Unmarshal(config); err != nil {
+		return 0, nil, nil, err
+	}
+
+	// Get the change set for the current configuration, again not bothering to
+	// create if it doesn't exist
+	csKey := fmtChangeSetKey(m.key, configVal.Version())
+	csVal, err := m.kv.Get(csKey)
+	if err != nil {
+		if err == kv.ErrNotFound {
+			// It's ok, just means no pending changes
+			return configVal.Version(), config, nil, nil
+		}
+
+		return 0, nil, nil, err
+	}
+
+	var changeset changesetpb.ChangeSet
+	if err := csVal.Unmarshal(&changeset); err != nil {
+		return 0, nil, nil, err
+	}
+
+	if changeset.State == changesetpb.ChangeSetState_COMMITTED {
+		// There are no pending changes
+		return configVal.Version(), config, nil, nil
+	}
+
+	// Retrieve the changes
+	changes := proto.Clone(m.changesType)
+	if err := proto.Unmarshal(changeset.Changes, changes); err != nil {
+		return 0, nil, nil, err
+	}
+
+	return configVal.Version(), config, changes, nil
+}
+
+func (m manager) Commit(version int, apply ApplyFn) error {
+	// Get the current configuration, but don't bother trying to create it if it doesn't exist
+	configVal, err := m.kv.Get(m.key)
+	if err != nil {
+		return err
+	}
+
+	// Confirm the version does exist...
+	if configVal.Version() < version {
+		return ErrUnknownVersion
+	}
+
+	// ...and that it hasn't already been committed
+	if configVal.Version() > version {
+		return ErrAlreadyCommitted
+	}
+
+	// Retrieve the config data so that we can transform it appropriately
+	config := proto.Clone(m.configType)
+	if err := configVal.Unmarshal(config); err != nil {
+		return err
+	}
+
+	// Get the change set for the current configuration, again not bothering to create
+	// if it doesn't exist
+	csKey := fmtChangeSetKey(m.key, configVal.Version())
+	csVal, err := m.kv.Get(csKey)
+	if err != nil {
+		return err
+	}
+
+	var changeset changesetpb.ChangeSet
+	if err := csVal.Unmarshal(&changeset); err != nil {
+		return err
+	}
+
+	// If the change list is already committed, bail out
+	if changeset.State == changesetpb.ChangeSetState_COMMITTED {
```

Comment:
Good catch.  We still want to allow commits against a COMMITTING ChangeSet to proceed, otherwise transient failures (such as network errors) that occur during Commit() will render the ChangeSet unusable without manual repair. But we can avoid this particular race condition by just not updating the ChangeSet if its state is already COMMITTING.